src/profile-logic/mozilla-symbolication-api.js

```diff
@@ -0,0 +1,181 @@
+        throw new Error('Expected `stack` to be an array');
+      }
+      for (const frameInfo of stack) {
+        if (!(frameInfo instanceof Object)) {
+          throw new Error('Expected `frameInfo` to be an object');
+        }
+        if (
+          !('module_offset' in frameInfo) ||
+          !('module' in frameInfo) ||
+          !('frame' in frameInfo)
+        ) {
+          throw new Error(
+            'Expected frameInfo to have `module_offset`, `module` and `frame` properties'
+          );
+        }
+      }
+    }
+  }
+  return result;
+}
+
+// Request symbols for the given addresses and libraries using the Mozilla
+// symbolication API.
+// Returns an array of promises, one promise per LibSymbolicationRequest in
+// requests. If the server does not have symbol information for a given library,
+// the promise for that library will fail.
+// That's the reason why this function does not return just one promise: We want
+// to indicate failure status for each library independently. Under the hood,
+// only one request is made to the server.
+export function requestSymbols(
+  requests: LibSymbolicationRequest[]
+): Array<Promise<Map<number, AddressResult>>> {
+  const addressArrays = requests.map(({ addresses }) => Array.from(addresses));
+  const body = {
+    memoryMap: requests.map(({ lib: { debugName, breakpadId } }) => [
+      debugName,
+      breakpadId,
+    ]),
+    stacks: addressArrays.map((addressArray, libIndex) =>
+      addressArray.map(addr => [libIndex, addr])
+    ),
+  };
+
+  const jsonPromise = fetch('https://symbols.mozilla.org/symbolicate/v5', {
+    body: JSON.stringify(body),
+    method: 'POST',
+    mode: 'cors',
+  }).then(response => response.json());
+
+  return requests.map(async function(request, libIndex) {
+    const { lib } = request;
+    const { debugName, breakpadId } = lib;
+
+    let json;
+    try {
+      json = _ensureIsAPIResult(await jsonPromise).results[0];
+    } catch (error) {
+      throw new SymbolsNotFoundError(
+        'There was a problem with the JSON returned by the symbolication API.',
+        lib,
+        error
+      );
+    }
+
+    if (!json.found_modules[`${debugName}/${breakpadId}`]) {
+      throw new SymbolsNotFoundError(
+        `The symbol server does not have symbols for ${debugName}/${breakpadId}.`,
+        lib
+      );
+    }
+
+    const addressInfo = json.stacks[libIndex];
+    const addressArray = addressArrays[libIndex];
+    if (addressInfo.length !== addressArray.length) {
+      throw new SymbolsNotFoundError(
+        'The result from the symbol server has an unexpected length.',
+        lib
+      );
+    }
+
+    const results = new Map();
+    for (let i = 0; i < addressInfo.length; i++) {
+      const address = addressArray[i];
+      const info = addressInfo[i];
+      if (
+        'function' in info &&
+        info.function !== undefined &&
+        'function_offset' in info &&
+        info.function_offset !== undefined
```

Comment:
nit: why don't you just use `info.function !== undefined && info.function_offset !== undefined` ?

Reply:
No reason. I think Firefox's JavaScript strict mode logs warnings when accessing a field that doesn't exist, but I should probably ignore the fear that those warnings instilled in me.